sdk/eventgrid/eventgrid/src/eventGridClient.ts

```diff
@@ -1,106 +1,185 @@
-   * @param events An array of events to be published to Event Grid.
-   * @param options The optional parameters
-   * @param callback The callback
+   * Publishes events written using a custom schema. The topic must be configured to expect events in a custom schema.
+   *
+   * @param message One or more events to publish
    */
-  publishEvents(topicHostname: string, events: Models.EventGridEvent[], options: msRest.RequestOptionsBase, callback: msRest.ServiceCallback<void>): void;
-  publishEvents(topicHostname: string, events: Models.EventGridEvent[], options?: msRest.RequestOptionsBase | msRest.ServiceCallback<void>, callback?: msRest.ServiceCallback<void>): Promise<msRest.RestResponse> {
-    return this.sendOperationRequest(
-      {
-        topicHostname,
-        events,
-        options
-      },
-      publishEventsOperationSpec,
-      callback);
+  sendCustomSchemaEvents(
+    events: Record<string, any>[],
+    options?: SendCustomSchemaEventsOptions
+  ): Promise<RestResponse> {
+    return this.client.publishCustomEventEvents(this.endpointUrl, events, options);
   }
-}
 
-// Operation Specifications
-const serializer = new msRest.Serializer(Mappers);
-const publishEventsOperationSpec: msRest.OperationSpec = {
-  httpMethod: "POST",
-  path: "api/events",
-  urlParameters: [
-    Parameters.topicHostname
-  ],
-  queryParameters: [
-    Parameters.apiVersion
-  ],
-  headerParameters: [
-    Parameters.acceptLanguage
-  ],
-  requestBody: {
-    parameterPath: "events",
-    mapper: {
-      required: true,
-      serializedName: "events",
-      type: {
-        name: "Sequence",
-        element: {
-          type: {
-            name: "Composite",
-            className: "EventGridEvent"
-          }
-        }
-      }
-    }
-  },
-  responses: {
-    200: {},
-    default: {
-      bodyMapper: Mappers.CloudError
+  /**
+   * Generate a shared access signature, which allows a client to send events to an Event Grid Topic for a limited period of time. This
+   * function may only be called when the EventGridClient was constructed with a KeyCredential instance.
+   *
+   * @param expiresOn The time at which the shared signature is no longer valid.
+   */
+  generateSharedAccessSignature(expiresOnUtc: Date): string {
+    if (!this.keyCredential) {
+      throw new Error(
+        "generateSharedAccessSignature may only be called when the client is constructed with a key credential"
+      );
     }
-  },
-  serializer
-};
-
-export {
-  EventGridClient,
-  EventGridClientContext,
-  Models as EventGridModels,
-  Mappers as EventGridMappers
-};
+
+    const expiresOnString = dateToServiceTimeString(expiresOnUtc);
+    const unsignedSas = `r=${encodeURIComponent(
+      `${this.endpointUrl}?apiVersion=${this.apiVersion}`
+    )}&e=${encodeURIComponent(expiresOnString)}`;
+    const digest = createHmac("sha256", Buffer.from(this.keyCredential?.key, "base64"))
```

Comment:
This client doesn't support the browser at the moment, right? Even so, it might be useful to extract node-specific pieces into separate files so you don't have `Buffer` in your main client class.

Reply:
> This client doesn't support the browser at the moment, right?

It actually does support the browser, but this specifically was the one thing that was unsupported (and I have a guard in the tests so we don't test this method in the browser today).

I want to add browser support and I think I can follow a similar strategy that Richard followed recently to add support for doing HMACs in the browser.